pyteal/ast/router.py

```diff
@@ -0,0 +1,559 @@
+    )
+    on_call: Optional[Expr | SubroutineFnWrapper | ABIReturnSubroutine] = field(
+        kw_only=True, default=None
+    )
+
+    @staticmethod
+    def never() -> "OCAction":
+        return OCAction()
+
+    @staticmethod
+    def create_only(
+        f: Expr | SubroutineFnWrapper | ABIReturnSubroutine,
+    ) -> "OCAction":
+        return OCAction(on_create=f)
+
+    @staticmethod
+    def call_only(
+        f: Expr | SubroutineFnWrapper | ABIReturnSubroutine,
+    ) -> "OCAction":
+        return OCAction(on_call=f)
+
+    @staticmethod
+    def always(
+        f: Expr | SubroutineFnWrapper | ABIReturnSubroutine,
+    ) -> "OCAction":
+        return OCAction(on_create=f, on_call=f)
+
+
+OCAction.__module__ = "pyteal"
+
+
+@dataclass(frozen=True)
+class OCActions:
+    """
+    OnCompletion Actions keep track of bare-call registrations to all OnCompletion cases.
+    """
+
+    close_out: OCAction = field(kw_only=True, default=OCAction.never())
+    clear_state: OCAction = field(kw_only=True, default=OCAction.never())
+    delete_application: OCAction = field(kw_only=True, default=OCAction.never())
+    no_op: OCAction = field(kw_only=True, default=OCAction.never())
+    opt_in: OCAction = field(kw_only=True, default=OCAction.never())
+    update_application: OCAction = field(kw_only=True, default=OCAction.never())
+
+    def dictify(self) -> dict[EnumInt, OCAction]:
+        return {
+            OnComplete.CloseOut: self.close_out,
+            OnComplete.ClearState: self.clear_state,
+            OnComplete.DeleteApplication: self.delete_application,
+            OnComplete.NoOp: self.no_op,
+            OnComplete.OptIn: self.opt_in,
+            OnComplete.UpdateApplication: self.update_application,
+        }
+
+    def is_empty(self) -> bool:
+        for oc_action in self.dictify().values():
+            if oc_action.on_call is not None or oc_action.on_create is not None:
+                return False
+        return True
+
+
+OCActions.__module__ = "pyteal"
+
+
+@dataclass(frozen=True)
+class CondNode:
+    condition: Expr
+    branch: Expr
+
+
+CondNode.__module__ = "pyteal"
+
+
+@dataclass
+class CategorizedCondNodes:
+    method_calls_create: list[CondNode] = field(default_factory=list)
+    bare_calls_create: list[CondNode] = field(default_factory=list)
+    method_calls: list[CondNode] = field(default_factory=list)
+    bare_calls: list[CondNode] = field(default_factory=list)
+
+    def program_construction(self) -> Expr:
+        concatenated_ast = (
+            self.method_calls_create
+            + self.bare_calls_create
+            + self.method_calls
+            + self.bare_calls
+        )
+        if len(concatenated_ast) == 0:
+            raise TealInputError("ABIRouter: Cannot build program with an empty AST")
+        program: Cond = Cond(*[[n.condition, n.branch] for n in concatenated_ast])
+        return program
+
+
+CategorizedCondNodes.__module__ = "pyteal"
+
+
+class Router:
```

Comment:
We can put this under `router`, or at the head of this file, see how it fits. Since I felt like the whole file of stuffs are all quite experimental...

Reply:
The latest commit contains a disclaimer on the top of `router.py`, feel free to change the wording and contents.

Reply:
@ahangsu Thanks for making an attempt.  Resolving - I opened new threads with additional feedback.